runtime/parachains/src/scheduler.rs

```diff
@@ -2038,4 +2051,119 @@ mod tests {
 			}
 		});
 	}
+
+	#[test]
+	fn no_panic_without_parathread_cores() {
+		let genesis_config = MockGenesisConfig {
+			configuration: crate::configuration::GenesisConfig {
+				config: no_parathread_cores_config(),
+				..Default::default()
+			},
+			..Default::default()
+		};
+
+		let chain_a = ParaId::from(1);
+		let chain_b = ParaId::from(2);
+
+		let schedule_blank_para = |id, is_chain| Paras::schedule_para_initialize(id, ParaGenesisArgs {
+			genesis_head: Vec::new().into(),
+			validation_code: Vec::new().into(),
+			parachain: is_chain,
+		});
+
+		new_test_ext(genesis_config).execute_with(|| {
+			// register 2 parachains
+			schedule_blank_para(chain_a, true);
+			schedule_blank_para(chain_b, true);
+
+			// start a new session to activate, 5 validators for 5 cores.
+			run_to_block(1, |number| match number {
+				1 => Some(SessionChangeNotification {
+					new_config: no_parathread_cores_config(),
+					validators: vec![
+						ValidatorId::from(Sr25519Keyring::Alice.public()),
+						ValidatorId::from(Sr25519Keyring::Bob.public()),
+						ValidatorId::from(Sr25519Keyring::Charlie.public()),
+						ValidatorId::from(Sr25519Keyring::Dave.public()),
+						ValidatorId::from(Sr25519Keyring::Eve.public()),
+					],
+					..Default::default()
+				}),
+				_ => None,
+			});
+
+			run_to_block(2, |_| None);
+
+			assert_eq!(Scheduler::scheduled().len(), 2);
+
+			// cores 0, 1, and 2 should be occupied. mark them as such.
+			Scheduler::occupied(&[CoreIndex(0), CoreIndex(1)]);
+
+			{
+				let cores = AvailabilityCores::get();
+
+				assert!(cores[0].is_some());
+				assert!(cores[1].is_some());
+
+				assert!(Scheduler::scheduled().is_empty());
+			}
+
+			run_to_block(3, |_| None);
+
+			// now note that cores 0 and 1 were freed.
+			Scheduler::schedule(vec![
+				(CoreIndex(0), FreedReason::Concluded),
+				(CoreIndex(1), FreedReason::Concluded),
+			]);
+
+			// clear up chain_a
+			Paras::schedule_para_cleanup(chain_a);
+			assert_eq!(Paras::outgoing_paras(), vec![chain_a]);
+			assert_eq!(AvailabilityCores::get().len(), 2);
+
+			run_to_block(4, |number| match number {
+				4 => Some(SessionChangeNotification {
+					new_config: no_parathread_cores_config(),
+					validators: vec![
+						ValidatorId::from(Sr25519Keyring::Alice.public()),
+						ValidatorId::from(Sr25519Keyring::Bob.public()),
+						ValidatorId::from(Sr25519Keyring::Charlie.public()),
+						ValidatorId::from(Sr25519Keyring::Dave.public()),
+						ValidatorId::from(Sr25519Keyring::Eve.public()),
+					],
+					..Default::default()
+				}),
+				_ => None,
+			});
+
+			assert_eq!(Paras::parachains(), vec![chain_b]);
+
+			{
+				assert_eq!(AvailabilityCores::get().len(), 1);
+
+				let scheduled = Scheduler::scheduled();
+
+				assert_eq!(scheduled.len(), 2);
```

Comment:
Why should `scheduled` have 2 items, including `chain_a`, if `chain_a` has been de-registered?